StatsNavigation: Remove bad plugins queries

sirreal opened the pull request:
While working on the Activity Log, I noticed bad plugin queries being fired to `/sites/undefined/plugins` on every request.

This PR addresses the problem by correcting the logic and not mounting the query component when a `siteId` is missing.

The PR also improves the query component's `propTypes` to better validate the array of `siteIds` which should help prevent future problems.

The issue appears to have been introduced in 6488094ccb. I'm curious about the motivation for querying plugins from `StatsNavigation` cc/ @psealock. It seems we should either do a better job of querying or move this query component to a different location.

## Testing
* URLs which exhibited the bug are `/stats/insights` and `/stats/activity` for Jetpack sites
* In production, you would see a bad network request to `/sites/undefined/plugins`.
* Visit https://calypso.live/stats/insights?branch=fix/sites-undefined-plugins-request and ensure you _do not_ see the bad request.
* Revert the change from a780648 to see the `propTypes` warning if bad props are passed to the query component.
* Ensure you don't see the store toggle on any of the global stats pages (day, month, year at `/stats` with no site selected).
* Ensure that using the store toggle is visibile works well from all the sections at `/stats/$SECTION/$SITE`. Insights and Activity should both navigate to Day on the store, as they don't have a corresponding seciton.

sirreal commented:
More context:

`siteId` could be added [here](https://github.com/Automattic/wp-calypso/blob/a229ad1743227fb42d5beae576e9fd663a42281d/client/my-sites/stats/activity-log/index.jsx#L336-L340) to ensure this query fires correctly, but plugins don't seem to make sense. That's why I'm hoping to understand the original motivations @psealock 🙂

psealock commented:
>siteId could be added here to ensure this query fires correctly, but plugins don't seem to be relevant there. That's why I'm hoping to understand the original motivations

When a user has JetPack and WooCommerce installed for a particular site, we render this pill to toggle between Site Stats and Store Stats.

![screen shot 2017-08-10 at 9 25 44 am](https://user-images.githubusercontent.com/1922453/29144797-6b10cd6a-7dae-11e7-8e0a-ecba6852f160.png)

So `StatsNavigation` needs to ensure the plugins are being queried. At the very least, my error was not making `siteId` a required prop along with `isJetpack` and `isStore`. Since those props are the responsibility of the parent, `QueryJetpackPlugins` should probably also be handled by the parent.

Alternatively, `StatsNavigation` can accept `siteId` and handle the querying and selectors `isJetpack` and `isStore`.

psealock commented:
@sirreal Thanks for bringing this to my attention. I made changes here https://github.com/Automattic/wp-calypso/pull/17063/commits/8842905103d74b9481403007518861709066a0a3

* `< StatsNavigation />` now is responsible for querying plugins, determining `isJetpack` and `isStore` independently of its parent.
* `< StatsNavigation />` now accepts `siteId` as an optional parameter. If no `siteId` is supplied, `isJetpack`, `isStore` are false so that Activity Log and Site/Store pill are not rendered. This is useful for the Overview Page (`/stats`).
* Made corresponding changes to "Insights" and "Activity Log" pages.

Is this a good solution? Let me know if there are any problems and I can address.

sirreal commented:
Thanks for looking at this!

I rebased and added some additional (mostly trivial) cleanup.

We're expecting to and should receive a `siteId`, so I think it makes sense to mark it as required (done in 4612394fe9e83167d17c99f9b4c52b07df6ecd8b).

I do have a few concerns. The toggle works well on the sections that correspond to the store sections, but Insights and Activity do not correspond. Using the toggle on either of those pages is significantly slower and seems to possible flash a 404 page?

I think it makes sense to try to better handle this routing, probably just sending `insights` and `activity` to the store `day`. It might be worth extracting the `SegmentedToggle` into it's own component as it grows in complexity.

Also when toggling from Activity and Insights, something happens with spacing on the right of the nav:

![nav](https://user-images.githubusercontent.com/841763/29164377-77998bb0-7dbf-11e7-81da-a3074a8e9127.gif)

I spent a bit looking at this issue but haven't come up with anything yet.

These are things that are being introduced by this PR so it might make sense to fix them here before it's merged.

psealock commented:
@sirreal Good catch on the mismatched sections. I made a few changes https://github.com/Automattic/wp-calypso/pull/17063/commits/783d5b067b1ba7e357be5bb73f8e7d0f66030359

* `StatsNavigation`'s `siteId` propType is no longer `isRequired`. The Overview page (`http://calypso.localhost:3000/stats`) has the component, but since no site is selected, it can't pass along an id. This results in "Insights" and "Activity Log" not being rendered, which makes sense.
* Import constants from Store Stats and compare `section` to make sure Store Stats has the corresponding tab, otherwise default to 'day' as you suggested. 
* The 404 and CSS error is taken care of by this^^^ change.

sirreal commented:
> I also ran Prettier on StatsNavigation

❤️ 

@psealock Very nice, everything LGTM. I think we're done here 😀 

Let's get a review and :shipit: !

mcsf commented on client/my-sites/stats/stats-navigation/index.jsx:
```diff
@@ -13,9 +17,12 @@ import NavItem from 'components/section-nav/item';
 import QueryJetpackPlugins from 'components/data/query-jetpack-plugins';
+import { isPluginActive } from 'state/selectors';
+import { isJetpackSite } from 'state/sites/selectors';
+import { UNITS as StoreStatsTabs } from 'extensions/woocommerce/app/store-stats/constants';
```
@sirreal: Is it intentional that a file in Calypso's core is requiring an extension? Is there a plan to eliminate this? I don't think it's a precedent we want to set.

FWIW there is prior art for _component decoration_ in Calypso's first concept extensions PR:

https://github.com/Automattic/wp-calypso/pull/8103/files

^ a search for `decorat` is a good starting point.

sirreal replied:
Thanks @mcsf, I didn't notice that. Looks like something we'll need to clean up.

/cc @psealock

psealock replied:
Thanks @mcsf 

>I don't think it's a precedent we want to set.

I think you are absolutely right. The import is using the tabs in Store Stats as a filter to make sure a link exists on the Store Stats side
```js
[ 'day', 'week', 'month', 'year' ]
```
There are discussions happening about changing Stats Navigation that will make this check obsolete, https://github.com/Automattic/wp-calypso/issues/17291.

I'll open a separate PR to address the bad pattern in the meantime.

psealock replied:
Continuing PR here https://github.com/Automattic/wp-calypso/pull/17440
